### PR TITLE
cheroot #42 -- remove unittest.testcase and ddt dependencies from unit tests

### PR DESCRIPTION
**Pull request by jeffvanvoorst**
* **What kind of change does this PR introduce?**
  - [ ] bug fix
  - [ ] feature
  - [ ] docs update
  - [X] tests/coverage improvement
  - [X] refactoring
  - [ ] other

* **What is the related issue number (starting with `#`)**
#42

* **What is the current behavior?** (You can also link to an open issue here)
Tests rely on unittest.testcase and use `ddt`

* **What is the new behavior (if this is a feature change)?**
Remove dependencies on `ddt` and `unittest.testcase`

* **Other information**:
So far, I haven't been able to use the WebTest.TestApp (or maybe it is wsgiproxy.HostProxy) without having a WSGI app loaded.  I am either calling an invalid URL or the server is accessed differently than the previous test code.  The exception is basically that "None doesn't have a start_response".

Need to remove stylistic differences and cleanup files once the changes are agreeable to the cheroot maintainers.

I am not sure how to run the code through CI/automated processes -- maybe I need a branch directly from the cheroot master?

* **Checklist**:

  - [X] I think the code is well written
  - [ ] I wrote [good commit messages][1]
  - [ ] I have [squashed related commits together][2] after the changes have been approved
  - [X] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [ ] I used the same coding conventions as the rest of the project
  - [ ] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [ ] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/116)

### CODE REVIEW

File: cheroot/test/conftest.py [hunk 1]
```diff
@@ -25,3 +30,76 @@ def wsgi_server_client(wsgi_server):
 def native_server_client(native_server):
     """Create a test client out of given HTTP server."""
     return get_server_client(native_server)
+
+
+@pytest.fixture
+def make_session_for_cheroot(wsgi_server):
+    '''
```
Comment (webknjaz) [inline cheroot/test/conftest.py]: Use PEP257 for docstrings

File: cheroot/test/conftest.py [hunk 2]
```diff
@@ -25,3 +30,76 @@ def wsgi_server_client(wsgi_server):
 def native_server_client(native_server):
     """Create a test client out of given HTTP server."""
     return get_server_client(native_server)
+
+
+@pytest.fixture
+def make_session_for_cheroot(wsgi_server):
```
Comment (webknjaz) [inline cheroot/test/conftest.py]: Can this be useful for end-users of cheroot in the testing of their apps? If yes, then it should go to `cheroot.testing` module.

File: cheroot/test/test_more.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import os
+import pytest
+import requests
+import ssl
+from cheroot.ssl.builtin import BuiltinSSLAdapter
+
+def ssl_file(filename):
```
Comment (webknjaz) [inline cheroot/test/test_more.py]: Can you use `trustme` for SSL certificates creation/testing?

Comment (jeffvanvoorst) [reply cheroot/test/test_more.py]: I think that is relatively straightforward to do.

Comment (webknjaz) [reply cheroot/test/test_more.py]: Yeah, please do

**Comment by webknjaz** [comment]
@jeffvanvoorst you can find linter errors in CI @ https://travis-ci.org/cherrypy/cheroot/jobs/431016556#L483
Also, you can run `tox -e pre-commit` for those locally.

**Comment by webknjaz** [comment]
Hint: you can enable tests in your fork here https://travis-ci.com/jeffvanvoorst/cheroot and then you'll be able to restart tests there by yourself if needed :)

File: .coveragerc [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# .coveragerc to control coverage.py
```
Comment (webknjaz) [inline .coveragerc]: We don't need this

File: cheroot/test/test_errors.py [hunk 5]
```diff
@@ -12,7 +13,7 @@
         (
             ('EPROTOTYPE', 'EAGAIN', 'EWOULDBLOCK',
              'WSAEWOULDBLOCK', 'EPIPE'),
-            [91, 11, 32]
+            [32, 10041, 11, 10035] if os.name == 'nt' else [91, 11, 32]
```
Comment (webknjaz) [inline cheroot/test/test_errors.py]: This must go to a separate PR.

Comment (webknjaz) [reply cheroot/test/test_errors.py]: (and use `system.platform()` instead)

File: cheroot/test/test_more.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    cert_type = request.param
+    cert_type_to_args = {
+        "client":            ["localhost", "127.0.0.1", "::1"],
+        "client_ip":         ["127.0.0.1", "::1"],
+        "client_wildcard":   ["*.localhost"],
+        "client_wrong_host": ["cherrypy.org", "github.com"],
+    }
+
+    client_cert = None
+    if cert_type == "client_wrong_ca":
```
Comment (webknjaz) [inline cheroot/test/test_more.py]: There's too much logic in the test case, it must go to separate tests, which should be as simple as possible.

File: cheroot/test/test_more.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    my_session = make_session_for_cheroot(
+        "localhost", prefix="https://", ssl_adapter=SSLAdapter, test_app_client="requests",
+        client_cert=client_cert_files.cert, client_key=client_cert_files.pkey)
+
+    # The pyriform requests adaptor eats the request verify and cert args as it isn't trivial to pass
+    # such args to an instant of WebTest.TestApp.
+    # Therefore, although the verify and cert args are set here, they are effectively ignored.
+    resp = my_session.get('https://localhost/', verify=SSLAdapter.certificate_chain,
+        cert=(client_cert_files.cert, client_cert_files.pkey))
+    if allowed is True:
```
Comment (webknjaz) [inline cheroot/test/test_more.py]: `is True` is redundant. Also, I don't like 2in1 tests

File: .coveragerc (hunk 4 above)
Comment (jeffvanvoorst) [reply .coveragerc]: not a problem -- my intention was to not have the unit test sources be considered when calculating code coverage.   However, if one looks more at the detailed list  and less at the over all number, then the information is approximately the same either way (implicitly versus explicitly ignoring the test files).

File: cheroot/test/test_more.py (hunk 7 above)
Comment (jeffvanvoorst) [reply cheroot/test/test_more.py]: Please explain the idea: 2in1 tests.  Is one parameterization enough?  
For example, could go very complex and test all client cert categories, 3 types of client certificate checking (from server point of view), and 2 servers from one test function.  
The other extreme is to have one test for each combination.
Which level of complexity/code sharing are you recommending?

File: cheroot/test/conftest.py (hunk 2 above)
Comment (webknjaz) [reply cheroot/test/conftest.py]: Well, `pyriform` is should help testing it without starting an HTTP server.

Comment (jeffvanvoorst) [reply cheroot/test/conftest.py]: My understanding is that this current fixture is for starting an HTTP server.

I had to replace certain parts of the TestApp and HostProxy in order to get this to work.  Also there is two sets of requests -- one used by HostProxy and one by Pyriform itself.

Comment (jeffvanvoorst) [reply cheroot/test/conftest.py]: Could probably think about how to add a Pyriform fixture for cherrypy.  Not sure how to add one for cheroot with a different framework.

Comment (webknjaz) [reply cheroot/test/conftest.py]: What do you mean by saying that? cheroot with a different framework?

File: cheroot/test/conftest.py [hunk 8]
```diff
@@ -25,3 +28,94 @@ def wsgi_server_client(wsgi_server):
 def native_server_client(native_server):
     """Create a test client out of given HTTP server."""
     return get_server_client(native_server)
+
+
+@pytest.fixture  # noqa: F811
```
Comment (webknjaz) [inline cheroot/test/conftest.py]: What do you ignore here?

File: cheroot/testing.py [hunk 9]
```diff
@@ -42,7 +52,8 @@ def cheroot_server(server_factory):
         try:
             actual_bind_addr = (interface, bind_port)
             httpserver = server_factory(  # create it
-                bind_addr=actual_bind_addr,
+                actual_bind_addr,
+                conf.pop('wsgi_app', None),
```
Comment (webknjaz) [inline cheroot/testing.py]: please use keyword args.

File: cheroot/testing.py [hunk 10]
```diff
@@ -21,10 +22,19 @@
 ANY_INTERFACE_IPV4 = '0.0.0.0'
 ANY_INTERFACE_IPV6 = '::'
 
+
+def my_crazy_app(environ, start_response):
```
Comment (webknjaz) [inline cheroot/testing.py]: I'm not sure that it belongs here. How can this be useful for testing apps by end-users?

File: cheroot/test/test_https.py [hunk 11]
```diff
@@ -2,146 +2,163 @@
...
+        # WebTest.TestApp.  Therefore, although the verify and cert args are
+        # set here, they are effectively ignored.
+        resp = my_session.get(
+            'https://localhost/',
+            verify=SSLAdapter.certificate_chain,
+            cert=(client_cert_files.cert, client_cert_files.pkey))
+        return resp
+
+    def expected_ssl_failure_status_code(self):
+        if six.PY3:
```
Comment (webknjaz) [inline cheroot/test/test_https.py]: Why is this different?

File: cheroot/test/test_https.py [hunk 12]
```diff
@@ -2,146 +2,163 @@
...
+    client_dir = tmpdir.mkdir('client')
+
+    client_cert_file = client_dir.join('client.cert')
+    client_cert_file.write(client_cert.cert_chain_pems[0].bytes())
+    client_pkey_file = client_dir.join('server_pkey')
+    client_pkey_file.write(client_cert.private_key_pem.bytes())
+    return CertKeyPair(str(client_cert_file), str(client_pkey_file), cert_type)
+
+
+class _TestClientCertValidation:
```
Comment (webknjaz) [inline cheroot/test/test_https.py]: Why do you need a class?

**Comment by jeffvanvoorst** [comment]
Not trying to be obstinate.  The part that I am not understanding is how to use the Pyriform adapter to a real cheroot server to help users.  My guess is that users would want to test their framework code and not the real webserver.  The proxy/test code will be different.

**Comment by jaraco** [comment]
I haven't spent enough time reviewing this to give it a meaningful review. The changes seem reasonable and genuine improvements. Can we say with confidence that this change won't break cherrypy's tests? If so, I'm almost inclined to accept it once @webknjaz is satisfied with the responses to his concerns.

**Comment by webknjaz** [comment]
@jeffvanvoorst this needs remaking because I've factored out the unittest and ddt stuff already so only testing facilitation left to do.
